File: AndroidBasics/notes/src/withGms/java/pan/alexander/notes/di/RoomModule.java
package pan.alexander.notes.di;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.room.Room;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import pan.alexander.notes.data.database.AppDatabase;
import pan.alexander.notes.data.database.TrashDao;

@Module
public class RoomModule {
    private final static String APP_DATABASE_NAME = "app-database";

    private final AppDatabase db;

    public RoomModule(@NonNull Application appContext) {
        db = Room.databaseBuilder(appContext, AppDatabase.class, APP_DATABASE_NAME)
                .fallbackToDestructiveMigration()
                .build();
    }

    @Provides
    @Singleton
    AppDatabase providesRoomDatabase() {
        return db;
    }

    @Provides
    @Singleton
    TrashDao providesTrashDao() {
        return db.getTrashDao();
    }
}

File: PopularLibs/githubclient/app/src/main/java/pan/alexander/githubclient/di/app/RoomModule.kt
package pan.alexander.githubclient.di.app

import android.content.Context
import androidx.room.Room
import dagger.Module
import dagger.Provides
import pan.alexander.githubclient.database.AppDatabase
import pan.alexander.githubclient.database.ListTypeConverter
import pan.alexander.githubclient.database.UserDao
import pan.alexander.githubclient.database.UserReposDao
import pan.alexander.githubclient.utils.configuration.ConfigurationManager
import javax.inject.Singleton

@Module
class RoomModule {
    @Provides
    @Singleton
    fun provideAppDatabase(
        context: Context,
        configurationManager: ConfigurationManager,
        listTypeConverter: ListTypeConverter
    ): AppDatabase = Room
        .databaseBuilder(
            context,
            AppDatabase::class.java,
            configurationManager.getAppDatabaseName()
        )
        .addTypeConverter(listTypeConverter)
        .fallbackToDestructiveMigration()
        .build()


    @Provides
    @Singleton
    fun provideUserDao(appDatabase: AppDatabase): UserDao = appDatabase.userDao

    @Provides
    @Singleton
    fun provideUserReposDao(appDatabase: AppDatabase): UserReposDao = appDatabase.userReposDao
}
